rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2168 @@
+# Safe Transmute RFC
+
+- Feature Name: `safe_transmute`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+We propose traits, namely `TransmuteInto` and `TransmuteFrom`, that are implemented *automatically* for combinations of types that may be safely transmuted. In other words, this RFC makes safe transmutation *as easy as 1..., 2..., `repr(C)`!*
+```rust
+use core::transmute::{
+    TransmuteInto,
+    stability::{PromiseTransmutableInto, PromiseTransmutableFrom},
+};
+
+#[derive(PromiseTransmutableInto, PromiseTransmutableFrom)]
+#[repr(C)]
+pub struct Foo(pub u8, pub u16);
+//                   ^ there's a padding byte here
+
+// Transmute fearlessly!
+let _ : Foo = 64u32.transmute_into(); // Alchemy Achieved!
+
+let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
+
+// error[E0277]: the trait bound `u32: TransmuteFrom<foo::Foo, _>` is not satisfied
+//   --> src/demo.rs:14:27
+//    |
+// 14 | let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
+//    |                           ^^^^^^^^^^^^^^ the trait `TransmuteFrom<foo::Foo, _>` is not implemented for `u32`
+//    |
+//   = note: required because of the requirements on the impl of `TransmuteInto<u32, _>` for `foo::Foo`
```

Comment:
Attempt to improve https://github.com/rust-lang/project-safe-transmute/pull/5#discussion_r464500451
```suggestion
use core::transmute::{
    TransmuteInto,
    stability::{PromiseTransmutableInto, PromiseTransmutableFrom},
};

#[derive(PromiseTransmutableInto, PromiseTransmutableFrom)] // declare `Foo` to be *stably* transmutable
#[repr(C)]
pub struct Foo(pub u8, pub u16);
//                    ^ there's a padding byte here, between these fields

// Transmute fearlessly!
let _ : Foo = 64u32.transmute_into(); // Alchemy Achieved!
//                  ^^^^^^^^^^^^^^ provided by the `TransmuteInto` trait

let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!

// error[E0277]: the trait bound `u32: TransmuteFrom<foo::Foo, _>` is not satisfied
//   --> src/demo.rs:15:27
//    |
// 15 | let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
//    |                           ^^^^^^^^^^^^^^ the trait `TransmuteFrom<foo::Foo, _>` is not implemented for `u32`
//    |
//   = note: required because of the requirements on the impl of `TransmuteInto<u32, _>` for `foo::Foo`
```